segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/stat/SegmentStatsRecorderTest.java

```diff
@@ -181,6 +181,31 @@ public void testMetrics() {
         assertEquals(NO_COUNTER_VALUE, getCounterValue(MetricsNames.SEGMENT_READ_BYTES, segmentName));
     }
 
+    @Test(timeout = 10000)
+    public void testTransactionMetrics() throws InterruptedException {
+        val segmentName = getStreamSegmentName();
+
+        long dataLength = 321;
+        int interval = 500;
+        val duration = Duration.ofMillis(2 * interval);
+        @Cleanup
+        val context = new TestContext(segmentName, duration, false);
+
+        context.statsRecorder.createSegment(segmentName, ScalingPolicy.ScaleType.BY_RATE_IN_KBYTES_PER_SEC.getValue(), 2, Duration.ofSeconds(1));
+        context.statsRecorder.recordAppend(segmentName, dataLength, 5, Duration.ofSeconds(2));
+        val txnName1 = NameUtils.getTransactionNameFromId(segmentName, UUID.randomUUID());
+        context.statsRecorder.createSegment(txnName1, ScalingPolicy.ScaleType.BY_RATE_IN_KBYTES_PER_SEC.getValue(), 2, Duration.ofSeconds(1));
+        Thread.sleep(duration.toMillis() / 2);
```

Comment:
From my understanding `Thread.sleep` only provides a lower bound on how much time will have elapsed before waking the thread, meaning it will not _absolutely guarantee_ that `segmentName` won't also be evicted alongside `txnName1` . Obviously `txnName1` can't actually be evicted because its prevented from entering the cache , but this test is setup to emulate the previous conditions and assert they cannot happen.

Reply:
You can use `TestUtils.await` or `AssertExtensions.assertEventuallyEquals`. If not already public, expose (via special getters) some info about your internal state (and mark it as `@VisibleForTesting`); then you can use one of those 2 methods to await some internal event to happen before resuming.